### PR TITLE
[WIP] More granular permission system

### PR DESCRIPTION
**Pull request by Salandora**
Alright, as mentioned in the IRC chat, I wanted to do a PR for this little project... Erm, huge project.

I tackled the more Granular Permission request (#1110).
It's not done yet, this here shall serve as a discussion platform, for a better overview of what has been implemented by now and what still needs to get implemented.

So to some informative stuff:
1. Implemented:
    - Permissions
    - Groups
    - Possibility to create new Groups through the API and therefore through the Web Interface
    - API checks for permissions
    - Web Interface checks for permissions
    - Plugins can define their own permissions
    - Admin and Guest groups are predefined non deletable groups
       - Admin group is static, admins will have all permissions
       - Guest group can be changed
    - Migration from old role system to new more granular permission system 
(still based on roles, but many more)

2. What is missing:
    - A bunch of useful automatic Tests
    - Some more hardening in the backend (Informations required)
      - How to protect the dataupdater?
      - How to protect the plugin messages?
      - Any way to generally protect the SockJS interface?
    - Some sort of fullscreen login (like the ForceLogin plugin) if a guest has absolutely no rights.
(Needs a bit of brainstoming)

Hopefully I haven't forgoten any important information.
Right now the only thing left for me is to wish everyone who wants to take a look at the code much fun and good luck :-)

Greetings
Salandora

### CODE REVIEW

File: src/octoprint/users.py [hunk 1]
```diff
@@ -1,647 +1,28 @@
...
-
-##~~ Apiuser object to use when global api key is used to access the API
-
-
-class ApiUser(User):
-	def __init__(self):
-		User.__init__(self, "_api", "", True, UserManager.valid_roles)
+		from octoprint.server import groupManager
+		groups = [groupManager.guests_group]
+		if "user" is None:
```
Comment (foosel) [inline src/octoprint/users.py]: I think you mean ``if "user" in roles``? Because that will never become true ;)

**Comment by foosel** [comment]
Just for the record:

> Some sort of fullscreen login (like the ForceLogin plugin) if a guest has absolutely no rights. (Needs a bit of brainstoming)

This at least should be covered as of 6d1f610. Currently it tests if the user has at least the `STATUS` and if not shows the dialog. We could of course also switch that to no permissions at all, but I figured, not having the `STATUS` permission would actually leave a pretty non-functional interface.

**Comment by foosel** [comment]
> How to protect the plugin messages?

How about changing `PluginManager.send_plugin_message(self, plugin, data)` into something like `PluginManager.send_plugin_message(self, plugin, data, permissions=None)`? That way plugins could provide additional limitations for messages they send, which the registered `PrinterStateConnection` instance (which I'm btw starting to think needs a rename 🤔) could then use to check against the authenticated user (about that, see [this gist](https://gist.github.com/foosel/68dfdbd9abeec4feb4a431d010c75bcf) of my initial idea how to accomplish that). Taking this a step further - we could adjust the `_emit` method to take additional permissions to check and have the other `on_*` methods adjusted accordingly (mostly to check against `Permissions.STATUS`).

Which brings me to think... do we maybe want to declare `STATUS` as something of a minimal limit and ALWAYS check plugin messages at least against that? Or do we maybe need something even more basic than this?

**Comment by foosel** [comment]
I did a first throw of what we talked about on IRC and here with regards to the socket lock down. Also added propagation of changes (with funny results - removing or adding the admin permission on a user now has the wrench pop up and out in a different browser). I still see some issues with the propagation of the group changes though, and a recent merge of devel broke some tests. I'll investigate that further.

**Comment by foosel** [comment]
At the risk of only talking to myself here ;) - Group changes now propagate properly, tests are fixed as well and I did a bit of refactoring to make the group/permission dialogs easier to take in. Some CSS stuff is still misbehaving though (of course they are).

**Comment by jneilliii** [comment]
How is this going to effect plugins and simpleapi mixin calls?

**Comment by foosel** [comment]
@jneilliii we'll make sure that `user_permission` and `admin_permission` stay there (albeit with a deprecation warning), and map to the new admin and user groups. So if you use the old mechanisms anywhere in your plugin to restrict APIs of any kind (simple or blueprint) to specific classes of users, that will continue to work. Still, long term goal should be that once 1.4.0 hits with the new permissions in place, plugins will utilize it.

**Comment by foosel** [comment]
Some visual changes. 

User list:

image

Group list:

image

User editor:

image

image

Also several bugs fixed again that crept in. I have the feeling that some permissions are not properly applied, switched a user from guest to user and the printer profile list and file list stayed empty. Also noticed:

  * [ ] API key generation in user manager works, but UI doesn't properly refresh. Probably some minor JS error.
  * [ ] Active flag changes so far do not log out users - this should be changed now that we also allow live changes when permissions change.
  * [ ] `user_permission` compatibility layer is still missing. Do we simply wrap a check of `not user.is_anonymous` here?
  * [ ] I'm still unsure whether we should enforce the "Users" group on all logged in users or not. Right now we don't.

**Comment by DongerZone** [comment]
Wondering if there is work to be done on this? I was about to spend my holidays working on this but saw its already in progress!

**Comment by foosel** [comment]
Nope, it's merged to `devel`

**Comment by DongerZone** [comment]
Exciting, can't wait to see it in effect!

**Pull request merged**